src/main.rs

```diff
@@ -60,33 +154,65 @@ fn main() {
     let parse_tree = parser::parse_prog(&mut token_iter);
 
     let mut code_gen = code_gen::CodeGen::new_with_config(config);
-    // Get the full original command.
-    code_gen.cmd = std::env::args().collect::<Vec<String>>().join(" ");
+     code_gen.cmd = std::env::args().collect::<Vec<String>>().join(" ");
 
-    // TODO(taegyunkim): Specify the root service name from commandline
-    code_gen.root_id = "productpage-v1";
-    code_gen.visit_prog(&parse_tree);
+     code_gen.root_id = matches.value_of("root_node").unwrap();
+     code_gen.visit_prog(&parse_tree);
 
-    // Open filter template.
-    let template_path = bin_dir.join("filter.cc.handlebars");
-    let display = template_path.display();
-    let mut template_file = match File::open(&template_path) {
-        Err(msg) => panic!("Failed to open {}: {}", display, msg),
-        Ok(file) => file,
-    };
 
-    let mut template_str = String::new();
-    match template_file.read_to_string(&mut template_str) {
-        Err(msg) => panic!("Failed to read {}: {}", display, msg),
-        Ok(_) => println!("Successfully read {}", display),
-    }
 
-    let handlebars = Handlebars::new();
+    // Use the information in the code generator code_gen and format it, using handlebars
+    // to a template with all the basic filter information enclosed
+    let c_mode = matches.value_of("compilation_mode").unwrap();
+    let filter_name = matches.value_of("filter_name").unwrap();
+    if c_mode == "sim" {
+        // Because we are making a library, not just one file, we need to copy over an example library.  Then,
+        // we have to write to three files:  Cargo.toml to edit the filter name, lib.rs to edit the filter name,
+        // and <filter_name>.rs for the actual filter implementation
+        
+        // Making new library folder
+        let mut lib_src_folder = filter_name.to_string();
```

Comment:
filter_name should probably be a Path already. `join` copies the string, `push` modifies it directly for `PathBuf`.

Reply:
If we make filter_name a path, then we we have the issue where we need the string version, both to create filter_name_types.rs, and as input into the handlebars formatting.  Are you suggesting just keeping around two versions, a string one and a path one?

Reply:
You can just convert the Path to string as needed with `file.to_str().unwrap()`. That should work. When is it actually needed as string?

Reply:
To make the filter, we create at least two files:  the filter itself, and filter_types.rs.  The second file lets the simulator know what type the filter function is and how to load it.  Because it is essentially like a .h file, I thought it would be appropriate to call it <filter_name>_types.rs.  So I've been making that, and the pathbuf library to my knowledge doesn't have anything that allows you to edit inside a path like that.  So since we need a string representation for the types file, I've been keeping it as a &str, but we could also rename that file to just types.rs if you think that's cleaner.

The handlebars comment was referencing figuring out the filter name in the Cargo.toml and lib.rs files, but given the earlier comment, that's no longer relevant.

So we could get around it being a &str, but I'd rather not because &str's make PathBuf declarations below easier to read because by defining them as
let filename : PathBuf = [filter_name, "src", "filter.rs"].iter().collect()
you can see the entire path and understand where it is in relation to other files.

Reply:
So I just tried to simplify this. I attached a patch. This only uses PathBuf manipulation. What do you think? There are some things I am confused about. For example, `let lib_src_folder: PathBuf = [output_name, "src"].iter().collect();` What kind of folder is the goal here? `filter/src`?
[file_path.patch.txt](https://github.com/dyn-tracing/tracing_compiler/files/5816815/file_path.patch.txt)

Reply:
That patch looks good.  Well at the very minimum we need two files produced by the compiler:  the filter itself and the types file.  So the most recent change I made was to put them into their own rust_filter folder, which also contains the Cargo.toml, lib.rs, graph_utils.rs files.  I think rust_filter should become the library, and then we can copy it over to the tracing_sim directory when we want to use it.  We have to put those auxiliary files somewhere, and I think it makes more logical sense that they are part of the compiler's work, and not the simulation's.

Reply:
Okay, sounds good to me. Feel free to merge after applying the patch. We can take care of minor issues that come up in the subsequent pull requests. Once I start automating the testing process for the cpp and the rs filter from the tracing env directory we will probably see where all the pain points are.

Reply:
Sounds like a plan!